src/main/java/com/google/sps/servlets/HandlePhotoShoppingServlet.java

```diff
@@ -0,0 +1,184 @@
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+
+import com.google.gson.Gson;
+
+import com.google.sps.GoogleShoppingQuerier;
+import com.google.sps.ShoppingQuerierConnectionException;
+import com.google.sps.data.Product;
+import com.google.sps.data.ShoppingQueryInput;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+
+/**
+ * When the user submits the form for image uploading, Blobstore processes the file upload and forwards
+ * the request to this servlet, which returns the product shopping results for the respective photo,
+ * in JSON format, along with the shopping query used to search.
+ */
+@WebServlet("handle-photo-shopping")
+public class HandlePhotoShoppingServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the BlobKey that points to the image uploaded by the user.
+    BlobKey uploadedImageBlobKey = getBlobKey(request, "photo");
+
+    // Send an error if the user did not upload a file.
+    if (uploadedImageBlobKey == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Client must upload an image file.");
+      return;
+    }
+
+    // Get the photo category (i.e. product, shopping-list or barcode) entered by the user.
+    if (request.getParameter("photo-category").isEmpty()) {
+      response.sendError(
+          HttpServletResponse.SC_BAD_REQUEST, "Client must select a photo category when submitting the form.");
+      return;
+    }
+    String photoCategory = request.getParameter("photo-category");
+
+    // Get the image the user uploaded as bytes.
+    byte[] uploadedImageBytes = getBlobBytes(uploadedImageBlobKey);
+    
+    // Call GoogleShoppingQuerier to return the extracted products data.
+    // First, build the shopping query input.
+    String shoppingQuery;
+    try {
+      shoppingQuery = getQuery(request.getParameter("photo-category"), uploadedImageBytes);
+    } catch(IllegalArgumentException exception) {
+      response.sendError(SC_INTERNAL_SERVER_ERROR, exception.getMessage());
+      return;
+    }
+    ShoppingQueryInput shoppingQueryInput = 
+        new ShoppingQueryInput.Builder(shoppingQuery).language("en").maxResultsNumber(24).build();
```

Comment:
I think the "resultsNum" should be one of the input in the original request. A default value will only be used when that field is not set in the request.

Reply:
Yes!! I also thought to ask the user to introduce the maximum results number, but I think there has been a misunderstanding, but I'll add it